### PR TITLE
Docs & Comments & Updates styles for EuiScreenReaderOnly

### PR DESCRIPTION
**Pull request by cchaos**
**tl;dr**

This still fixes nothing but the original issue of overflowing content in tables. At least the caveats are documented and we can take a few extra steps down the line to rectify some of them on our side.

---

# Docs

I re-wrote a lot of the Screen Reader Only docs to mention some of the nuances.

# Comments

I updated the mixin and added comments, more line-by-line, detailing why each was important.

# Updated styles

Based on your advice to just remove the mixin from the checkboxes and radios, I've adjusted those styles. 

Here are some screenshots of the placement of the input in all 3 browsers, Chrome - Safari - Firefox (which is blurry cuz i use it to test non-retina stuff). 

image
image

# EuiLink

I also added a style specifically to EuiLink that makes is `position: relative` if it contains that screen reader text which is only there if `target="_blank"`. This is just an example really, but we could/should comb through usages that we know exist inside a focusable element.

image

### CODE REVIEW

File: src/components/link/_link.scss [hunk 1]
```diff
@@ -1,6 +1,11 @@
 .euiLink {
   @include euiLink;
 
+  &[target='_blank'] {
+    // Make the screen reader only text positioned relatively against the link itself
+    position: relative;
+  }
+
```
Comment (constancecchen) [inline src/components/link/_link.scss:8]: [not a blocker, just a wild thought] For the external EuiLink use case, I'm wondering - why not skip the screen reader only text and just conditionally bake it into the `aria-label` on the external icon?



It looks like when we implemented the SR text it was a conscious decision to allow people to disable the icon on a `target="_blank"` link, but I wonder if it wouldn't be simpler to enforce the icon+SR text on new tab/window links (the same way we enforce other semantic/accessibility best practices).

Again, definitely not a blocker and happy to merge this PR w/o, just wanted to throw that out there to see what you thought!

Comment (cchaos) [reply src/components/link/_link.scss:8]: This was a direct implementation from the a11y-expert-no-longer-here. I'm guessing he had his reasons.

Comment (cchaos) [reply src/components/link/_link.scss:8]: Oh and it can go the other way too where the icon can be present without opening in a new tab.

File: src-docs/src/views/accessibility/accessibility_example.js [hunk 2]
```diff
@@ -75,40 +62,119 @@ export const AccessibilityExample = {
...
+      ],
+      text: (
+        <>
+          <h3>Within a button</h3>
+          <p>
+            If the screen reader content is <strong>contained within</strong> a
+            focusable element, you should consider adding{' '}
+            <EuiCode language="scss">{'position: relative;'}</EuiCode> to the
+            focusable element. This will fix any screen reader focus rings to
+            stay within the bounds of the focusable element.
```
Comment (constancecchen) [inline src-docs/src/views/accessibility/accessibility_example.js:101]: After some thought, I'm on the fence about adding an extra documentation section for this behavior, since as far as I can tell it mostly applies to just Safari + VO.

We might want to evaluate end-user usage and whether it's worth adding this documentation/advisement for a single browser+screenreader combo that WebAim reports in its 2021 survey to be [5% of desktop screen reader users](https://webaim.org/projects/screenreadersurvey9/#browsercombos) (compared to JAWS and NVDA on Windows, which is [90%+](https://webaim.org/projects/screenreadersurvey9/#os)). I think:

1. This is a non-issue for the vast majority screen reader users on desktop
    1. Although I do want to caveat that for mobile screen reader users (50%+ of SR users), iOS Safari + VO is 50% of the market share. However, VO doesn't behave the same way on mobile as it does on desktop and this may be a non-issue there - we won't know for sure until we test on an iOS device. But to that point: IIRC we don't technically support/test EUI or Kibana on mobile devices.
2. This is not an issue for Safari keyboard users, as Safari does not show the off-screen outline hitbox when not using VO
3. This is also purely a visual outline issue - for a tool that is meant for non-visual users. I acknowledge there's a spectrum of vision impairment and it's possible some screen reader users will still see the off-screen outline and may be slightly confused by it, but nevertheless the audio behavior still works perfectly. With that, I'm going to argue that it's not worth it to document a workaround for a problem that maybe half of 5% of our users might affect (and not functionally).

To clarify, I'm 100% on board with adding `position: relative` for our external EUI links since it's a fairly common pattern and a small/easy fix for us, I just think this extra documentation section for it is unnecessary since it's incredibly browser+screen reader specific and extremely low-impact, as above.

File: src-docs/src/views/accessibility/accessibility_example.js [hunk 3]
```diff
@@ -75,40 +62,119 @@ export const AccessibilityExample = {
...
+        EuiScreenReaderOnly,
+      },
+      snippet: `<EuiScreenReaderOnly showOnFocus>
+  <!-- visually hidden content, displayed on focus -->
+</EuiScreenReaderOnly>`,
+      demo: <ScreenReaderFocus />,
+    },
+    {
+      title: 'Skip link',
+      source: [
```
Comment (constancecchen) [inline src-docs/src/views/accessibility/accessibility_example.js:155]: I love the new page sections/organization!! It's a huge improvement and brings a ton of clarity to the previous iteration just IMO 😍

File: src/components/link/_link.scss (hunk 1 above)
Comment (constancecchen) [reply src/components/link/_link.scss:8]: Right yeah, the aria-label itself would need some conditional logic to detect `target="_blank"` and add the parenthetical, but that's fairly straightforward.

If you're not a huge fan of the concept of putting the SR text all on `aria-label` though, no worries, we can just leave it as-is 👍

**Comment by constancecchen** [comment]
Sorry, I definitely got sidetracked looking up screen reader usage and that ended up dominating my comments/thoughts, but I also wanted to say everything else in the PR looks super great! 3rd comment is definitely my only blocking-ish comment/discussion I'd like to have, otherwise I'm happy to merge this in to the open PR 🎉

File: src-docs/src/views/accessibility/accessibility_example.js (hunk 2 above)
Comment (cchaos) [reply src-docs/src/views/accessibility/accessibility_example.js:101]: I just err on giving the consumers full knowledge up front of known issues and how to work around them. If you feel strongly, feel free to remove.

File: src-docs/src/views/accessibility/accessibility_example.js [hunk 4]
```diff
@@ -51,19 +42,15 @@ export const AccessibilityExample = {
...
-            Use the <strong>EuiScreenReaderOnly</strong> component to visually
-            hide elements while still allowing them to be read by screen
-            readers. In certain cases, you may want to use the{' '}
-            <EuiCode>showOnFocus</EuiCode> prop to display screen reader-only
-            content when in focus.
+            Using <strong>EuiScreenReaderOnly</strong> hides the wrapped element
+            from the page, but keeps it accessible for screen readers to provide
+            more context. It should be used primarily to mask{' '}
+            <strong>text</strong> and requires the child to be a single React
+            element for cloning.
```
[the remark this answers is not in the document]
Comment (cchaos) [reply src-docs/src/views/accessibility/accessibility_example.js:53]: The detail can be helpful because it could make them spot issues if they have style conflicts between the added class and their element.

File: src-docs/src/views/accessibility/accessibility_example.js [hunk 5]
```diff
@@ -75,40 +62,119 @@ export const AccessibilityExample = {
               provides functionality or interactivity) is important enough to
               provide to screen reader users, it should probably be made
               available to all users.&quot;{' '}
-              <EuiLink
-                href="http://webaim.org/techniques/css/invisiblecontent/"
-                external
```
[the remark this answers is not in the document]
Comment (cchaos) [reply src-docs/src/views/accessibility/accessibility_example.js:80]: Yes, and no it's actually not valid because the indicator should only be used to indicate opening in a new tab.

[the remark this answers is not in the document]
Comment (constancecchen) [reply src-docs/src/views/accessibility/accessibility_example.js:80]: Ahhh!! TIL, thanks, that's super helpful to know!

[the remark this answers is not in the document]
Comment (constancecchen) [reply src-docs/src/views/accessibility/accessibility_example.js:80]: Also now I'm wondering if I've been using that incorrectly in Enterprise Search 😂 Thankfully I think most of our external links we set to `target="_blank"` anyway

**Comment by constancecchen** [comment]
Apologies for the slowness! I realized yesterday all my email notifications for this PR were going to my personal email address instead of my work inbox 🤦‍♀️ 

After some thought I still feel fairly strongly about not documenting the position relative workaround for Safari+VO (unless we have precedence elsewhere for browser-specific workarounds or issues in our documentation). I think what could make sense instead is to open a GitHub issue documenting the known behavior - this will make the issue discoverable for more GitHub users and also allow them to comment/discuss, which will give us more helpful insight into usage and users affected.

I'll go ahead and merge this PR, open an issue, and remove the section/example in the main branch if no objections!
